tests/unit/src/main/scala/tests/TestingServer.scala

```diff
@@ -862,6 +862,103 @@ final class TestingServer(
     }
   }
 
+  def assertTypeDefinition(
+      filenameStr: String = "",
+      queryStr: String,
+      expectedLocs: List[l.Location] = Nil,
+      root: AbsolutePath
+  ): Future[Unit] = {
+    val fMap =
+      FileLayout.mapFromString(queryStr).filter(_._1.endsWith(".scala"))
+
+    val (filename, query, uri) = if (filenameStr.isEmpty) {
+      val uriRegex = """^[^/]*/\*([^*]+)\*/[^/]*$""".r
+
+      fMap.find(_._2.contains("@@")) match {
+        case Some(file) =>
+          val (filename, content) = file
+          (
+            filename,
+            content.replaceAll("""/\*[^\*]*\*/""", ""),
+            content.lines.find(_.contains("@@")).collect {
+              case uriRegex(uri) => uri
+            }
+          )
+        case _ => (filenameStr, queryStr, None)
+      }
+    } else (filenameStr, queryStr, None)
+
+    val extractedExp = if (expectedLocs.isEmpty) {
```

Comment:
Can we move most of the logic from here to the TestSuite - it seems that it can be simplified a lot really.